### PR TITLE
Add docstrings and type annotations to `patroni/utils.py`

### PR DESCRIPTION
**Pull request by barthisrael**
References: PAT-43

### CODE REVIEW

File: patroni/utils.py [hunk 1]
```diff
@@ -123,10 +203,31 @@ def strtol(value, strict=True):
     return (None if strict else 1), value
 
 
-def strtod(value):
-    """As most as possible close equivalent of strtod(3) function used by postgres to parse parameter values.
-    >>> strtod(' A ') == (None, 'A')
-    True
+def strtod(value: Any) -> Tuple[float, str]:
```
Comment (CyberDem0n) [inline patroni/utils.py:206]:
```suggestion
def strtod(value: Any) -> Tuple[Union[float, None], str]:
```

File: patroni/utils.py [hunk 2]
```diff
@@ -216,14 +370,26 @@ def parse_int(value, base_unit=None):
...
 
-def parse_real(value, base_unit=None):
-    """
-    >>> parse_real(' +0.0005 ') == 0.0005
-    True
-    >>> parse_real('0.0005ms', 'ms') == 0.0
-    True
-    >>> parse_real('0.00051ms', 'ms') == 0.001
-    True
+def parse_real(value: Any, base_unit: Optional[str] = None) -> float:
```
Comment (CyberDem0n) [inline patroni/utils.py:373]:
```suggestion
def parse_real(value: Any, base_unit: Optional[str] = None) -> Union[float, None]:
```

File: patroni/utils.py [hunk 3]
```diff
@@ -258,30 +461,47 @@ def compare_values(vartype, unit, old_value, new_value):
...
-        :param delay: Initial delay between retry attempts.
-        :param backoff: Backoff multiplier between retry attempts. Defaults to 2 for exponential backoff.
-        :param max_jitter: Additional max jitter period to wait between retry attempts to avoid slamming the server.
-        :param max_delay: Maximum delay in seconds, regardless of other backoff settings. Defaults to one hour.
-        :param retry_exceptions: single exception or tuple"""
-
+    def __init__(self, max_tries: Optional[int] = 1, delay: Optional[float] = 0.1, backoff: Optional[int] = 2,
+                 max_jitter: Optional[float] = 0.8, max_delay: Optional[int] = 3600,
+                 sleep_func: Callable[[Union[int, float]], None] = _sleep, deadline: Union[int, float] = None,
+                 retry_exceptions: Union[Exception, Tuple[Exception]] = PatroniException) -> None:
```
Comment (CyberDem0n) [inline patroni/utils.py:493]:
```suggestion
                 sleep_func: Optional[Callable[[Union[int, float]], None]] = _sleep, deadline: Optional[Union[int, float]] = None,
                 retry_exceptions: Optional[Union[Exception, Tuple[Exception]]] = PatroniException) -> None:
```

File: patroni/utils.py [hunk 4]
```diff
@@ -295,36 +515,58 @@ def __init__(self, max_tries=1, delay=0.1, backoff=2, max_jitter=0.8, max_delay=
...
-    def copy(self):
-        """Return a clone of this retry manager"""
+    def copy(self) -> 'Retry':
+        """Return a clone of this retry manager."""
         return Retry(max_tries=self.max_tries, delay=self.delay, backoff=self.backoff,
                      max_jitter=self.max_jitter / 100.0, max_delay=self.max_delay, sleep_func=self.sleep_func,
                      deadline=self.deadline, retry_exceptions=self.retry_exceptions)
 
     @property
     def sleeptime(self):
```
Comment (CyberDem0n) [inline patroni/utils.py:530]:
```suggestion
    def sleeptime(self) -> float:
```

File: patroni/utils.py [hunk 5]
```diff
@@ -295,36 +515,58 @@ def __init__(self, max_tries=1, delay=0.1, backoff=2, max_jitter=0.8, max_delay=
...
+        """Get next cycle sleep time.
+
+        It is based on the current delay plus a number up to ``max_jitter``.
+        """
         return self._cur_delay + (random.randint(0, self.max_jitter) / 100.0)
 
     def update_delay(self):
```
Comment (CyberDem0n) [inline patroni/utils.py:537]:
```suggestion
    def update_delay(self) -> None:
```

File: patroni/utils.py [hunk 6]
```diff
@@ -295,36 +515,58 @@ def __init__(self, max_tries=1, delay=0.1, backoff=2, max_jitter=0.8, max_delay=
...
+        """Set next cycle delay.
+
+        It will be the minimum value between:
+            * current delay with ``backoff``; or
+            * ``max_delay``.
+        """
         self._cur_delay = min(self._cur_delay * self.backoff, self.max_delay)
 
     @property
     def stoptime(self):
```
Comment (CyberDem0n) [inline patroni/utils.py:547]:
```suggestion
    def stoptime(self) -> Union[float, None]:
```

File: patroni/utils.py [hunk 7]
```diff
@@ -295,36 +515,58 @@ def __init__(self, max_tries=1, delay=0.1, backoff=2, max_jitter=0.8, max_delay=
...
 
-    def __call__(self, func, *args, **kwargs):
-        """Call a function with arguments until it completes without throwing a `retry_exceptions`
-
-        :param func: Function to call
-        :param args: Positional arguments to call the function with
-        :params kwargs: Keyword arguments to call the function with
-
-        The function will be called until it doesn't throw one of the retryable exceptions"""
+    def __call__(self, func: Callable, *args: Any, **kwargs: Any) -> None:
```
Comment (CyberDem0n) [inline patroni/utils.py:551]:
```suggestion
    def __call__(self, func: Callable, *args: Any, **kwargs: Any) -> Any:
```

File: patroni/utils.py [hunk 8]
```diff
@@ -136,21 +237,57 @@ def strtod(value):
...
+        True
 
+        >>> rint(0.501) == 1
+        True
+
+        >>> rint(1.5) == 2
+        True
+    """
     ret = round(value)
     return 2.0 * round(value / 2.0) if abs(ret - value) == 0.5 else ret
```
[the remark this answers is not in the document]
Comment (barthisrael) [reply patroni/utils.py:259]: Thanks for the heads up! Based on Python documentation and on some tests, that seems the case indeed. I guess we can remove this function at this point.

[the remark this answers is not in the document]
Comment (barthisrael) [reply patroni/utils.py]: Just pushed a commit explaining why we are removing that function.

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**